src/Icicle/Common/Type.hs

```diff
@@ -43,24 +45,26 @@ import              P
 import qualified    Data.Map as Map
 
 
--- | Real values.
+-- | Types of real values.
 -- No functions here, because we don't want higher order functions in the generated code.
 -- This restriction should simplify code generation, because we won't need to
 -- deal with lambda lifting arbitrary functions.
-data ValType =
+data ValType' t =
```

Comment:
this is hell.
I have two different kinds of types: these, which are used in Core, but in Source I need more stuff like type variables, foralls, etc.
so I'm trying to "untie the recursive knot" here so I can reuse this for both cases, but it's just not particularly fun.
any suggestions of a better way to do this?